frontend/packages/operator-lifecycle-manager/src/components/create-operand.tsx

```diff
@@ -104,297 +178,465 @@ const defaultValueFor = (field: OperandField) => {
+      return {
+        ...formValuesAccumulator,
+        ...(value && { [field.path]: value }),
+      };
+    }, initialMetadataValues);
+  }, [fields, buffer]);
 
-  type FormValues = { [path: string]: any };
+  const [formValues, setFormValues] = React.useState<FormValues>(initialFormValues);
+  const [error, setError] = React.useState<string>();
+  const [formErrors, setFormErrors] = React.useState<FieldErrors>({});
 
-  const defaultFormValues = React.useMemo(
-    () =>
-      fields.reduce<FormValues>(
-        (allFields, field) => ({ ...allFields, [field.path]: defaultValueFor(field) }),
-        {},
+  // Group fields into advanced, arrayFieldGroup, fieldGroup, and normal fields for rendering.
+  // Note that arrayFieldGroup and fieldGroup fields are still flat after this. The memoized
+  // 'arrayFieldGroups' and 'fieldGroups' (below) further organizes these by their respective
+  // group descriptors to simplify rendering and state management.
+  const [advancedFields, arrayFields, groupFields, normalFields]: OperandField[][] = React.useMemo(
+    (): OperandField[][] =>
+      _.reduce(
+        fields,
+        (
+          [
+            advancedFieldsAccumulator = [],
+            arrayFieldsAccumulator = [],
+            groupFieldsAccumulator = [],
+            normalFieldsAccumulator = [],
+          ],
+          field,
+        ) => {
+          if (hasDescriptor(field, SpecCapability.arrayFieldGroup)) {
+            return [
+              advancedFieldsAccumulator,
+              [...arrayFieldsAccumulator, field],
+              groupFieldsAccumulator,
+              normalFieldsAccumulator,
+            ];
+          }
+          if (hasDescriptor(field, SpecCapability.fieldGroup)) {
+            return [
+              advancedFieldsAccumulator,
+              arrayFieldsAccumulator,
+              [...groupFieldsAccumulator, field],
+              normalFieldsAccumulator,
+            ];
+          }
+          if (hasDescriptor(field, SpecCapability.advanced)) {
+            return [
+              [...advancedFieldsAccumulator, field],
+              arrayFieldsAccumulator,
+              groupFieldsAccumulator,
+              normalFieldsAccumulator,
+            ];
+          }
+          return [
+            advancedFieldsAccumulator,
+            arrayFieldsAccumulator,
+            groupFieldsAccumulator,
+            [...normalFieldsAccumulator, field],
+          ];
+        },
+        [],
       ),
     [fields],
   );
 
-  const sampleFormValues = React.useMemo(
-    () =>
-      fields.reduce<FormValues>((allFields, field) => {
-        const sampleValue = _.get(sample, `spec.${field.path}`);
-        return sampleValue ? { ...allFields, [field.path]: sampleValue } : allFields;
-      }, {}),
-    [fields, sample],
-  );
+  // Create memoized arrayFieldGroups. Organizes arrayFields into a structure that is easy to map
+  // and render.
+  const arrayFieldGroups = React.useMemo(() => {
```

Comment:
Memoized arrayFieldGroups into a data structure that makes rendering and state updates much easier. This function manipulates a flat array of all fields belonging in arrayFieldGroups into a more easily iterable format:

```tsx
type arrayFieldGroups = {
  groupName: string;  // The name of the arrayFieldGroup as parsed from the spec descriptor
  fieldLists: OperandField[][] // OperandFields grouped by index
}[];
```